projects/frontend/src/draft-posts/index.md

```diff
@@ -0,0 +1,148 @@
+
+```go
+package main
+
+import (
+	"flag"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strconv"
+)
+
+func main() {
+	port := flag.Int("port", 8080, "port to listen on")
+	targetURL := flag.String("target-url", "", "downstream service url to proxy to")
+	flag.Parse()
+
+	u, err := url.Parse(*targetURL)
+	if err != nil {
+		log.Fatalf("Could not parse downstream url: %s", *targetURL)
+	}
+
+	proxy := httputil.NewSingleHostReverseProxy(u)
+
+	director := proxy.Director
+	proxy.Director = func(req *http.Request) {
+		director(req)
+		req.Header.Set("X-Forwarded-Host", req.Header.Get("Host"))
+		req.Host = req.URL.Host
+	}
+
+	http.HandleFunc("/", proxy.ServeHTTP)
+	log.Printf("Listening on port %d", *port)
+	log.Fatal(http.ListenAndServe(":"+strconv.Itoa(*port), nil))
+}
+```
+
+That’s it! This server can now handle proxying HTTP 1 and 2 as well as web socket connections. You’ll notice that I’ve configured the _proxy.Director_ field. The _ReverseProxy.Director_ is a function that modifies the incoming request before it is forwarded. The signature is as follows:
+
+```go
+Director func(*http.Request)
+```
+
+A common use case for the director function is altering request headers. There is a default director set by  _httputil.NewSingleHostReverseProxy_ . The default implementation takes care of setting the request scheme, host, and path. I found this useful and didn’t want to duplicate the code. I had to reset the _req.Host_  field to contact HTTPS endpoints . I’ve also included an example of setting a request header.
+
+## Capturing Metrics 
+Not too long ago I was asked to build a demo for one of our clients at work. The task was to capture the response metrics of a web service and add those metrics to [Prometheus](https://prometheus.io/) for display in [Grafana](https://grafana.com/). The only catch was that I did not have direct access to the web service and therefore couldn’t instrument the code. I completed the task by adapting the go reverse proxy to observe the network traffic coming from the service.
+
+Let’s extend our simple proxy to read and report metrics about the downstream service responses. To do this we’ll return to the  _httputil.ReverseProxy_ struct once more. It exposes a  struct field _ReverseProxy.ModifyResponse_ which gives us access to the http response before it goes back to the client.
+
+```go
+ModifyResponse func(*net/http.Response) error
+```
+
+HTTP bodies in go are implemented as _io.Reader_’s and therefore they can only be read once. If you would like to parse them before forwarding the request/response you will need to copy the body into a byte buffer and the reset the request/response body. An obvious drawback is that we buffer the entire response in memory without limit. This could lead to memory issues in production if you received a large response but for our use case this wasn’t an issue. Here’s a quick implementation to duplicate the response body.
```

Comment:
```suggestion
HTTP bodies in go are implemented as _io.Reader_s and therefore can only be read once. If you would like to parse them before forwarding the HTTP message you will need to copy the body into a byte buffer and the reset the body. An obvious drawback is that we buffer the entire response in memory without limit. This could lead to memory issues in production if you received a large response but for our use case this wasn’t an issue. Here’s a quick implementation to duplicate the response body.
```

Reply:
I would also specify at this point that there's no reason to actually read the body unless you actually need something out of it.